Fix network down up process

bjsowa opened the pull request:
Description
-----------
I couldn't find a lot of information about the process of network going down (only some of the ports even support it) but from what I understand, it should work like this:
1. When the port detects that the link is down, it informs the network stack by calling `FreeRTOS_NetworkDown` or by sending `eNetworkDownEvent` to IP task.
2. The network stack stops any ongoing DHCP or RA process, clears ARP, etc
3. The `pfInitialize` function is periodically polled by the network stack until it returns pdPASS which signals that the link went up again.
4. The DHCP/RA processes are restarted.

During port development I noticed 2 problems:
1. The DHCP/RA timers are not stopped when the network goes down. 
2. If all interfaces are up at some point, the network stack will not poll the `pfInitialize` function when an interface goes down as the `xAllNetworksUp` flag is set and network timer is not longer checked.

In this PR I fixed 1. by disabling the DHCP_RA timer for all endpoints when the network goes down. I also fixed 2. by removing `xAllNetworksUp` altogether as I didn't have any better idea. Let me know what you think.

Test Steps
-----------
Use a port that support network down event and try disconnecting it from the network and reconnecting it. If the port is implemented correctly, the DHCP process should be stopped when the link goes down and restarted when the link goes up again.

Checklist:
----------
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have tested my changes. No regression in existing tests.
- [x] I have modified and/or added unit-tests to cover the code changes in this Pull Request.


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

tony-josi-aws commented on source/FreeRTOS_IP_Timers.c:
```diff
@@ -55,16 +55,6 @@
-/* MISRA Ref 8.9.1 [File scoped variables] */
-/* More details at: https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/blob/main/MISRA.md#rule-89 */
-/* coverity[misra_c_2012_rule_8_9_violation] */
-static BaseType_t xAllNetworksUp = pdFALSE;
```
Instead of removing `xAllNetworksUp` maybe we could have a set/reset API to update the state of the variable when the network goes down maybe from `prvProcessNetworkDownEvent`.
This could avoid unnecesary looping through the interfaces whenever the network timer expires.

bjsowa replied:
I left `xAllNetworksUp` intact and added a Private API for setting it to pdFALSE.

moninom1 replied:
Yeah seems fine. You might have to add Unit Test for the new function you have added for code coverage.

htibosch commented on source/FreeRTOS_IP_Utils.c:
```diff
@@ -861,6 +861,22 @@ void prvProcessNetworkDownEvent( struct xNetworkInterface * pxInterface )
+            if( END_POINT_USES_DHCP( pxEndPoint ) )
+            {
+                /* Stop the DHCP process for this end-point. */
+                vIPSetDHCP_RATimerEnableState( pxEndPoint, pdFALSE );
```
When you try to end a DHCP procedure just by disabling the timer, there is a risk that the DHCP socket won't be closed.
In DHCP/IPv4, the socket has a reference count called `xDHCPSocketUserCount`. It should be closed by calling `prvCloseDHCPSocket`.
For DHCPv6, this is much the same, the socket is called `xDHCPv6Socket`, and it is protected by `xDHCPv6SocketUserCount`, and closed by calling `prvCloseDHCPv6Socket()`.

bjsowa replied:
I added `vDHCPStop` and `vDHCPv6Stop` functions that, besides disabling the timer, also call `prvCloseDHCPSocket` to attempt to close the socket. 

I also noticed that IPv4 implementation of DHCP is not aware which endpoints opened the socket and calling `prvCloseDHCPSocket` will decrease `xDHCPSocketUserCount` even when the endpoint did not open it, whereas the DHCPv6 implementation stores the socket locally for each endpoint. That's why I decided to change DHCPv4 implementation to be more in line with DHCPv6.

moninom1 commented on source/FreeRTOS_IP_Timers.c:
```diff
@@ -604,3 +604,13 @@ void vIPSetARPResolutionTimerEnableState( BaseType_t xEnableState )
+ *        Whenever the timer expires, all interfaces that are down will get a new NetworkDown
+ *        event.
+ */
+void vSetNotAllNetworksUp( void )
```
Can you please update the function name to be more clearer.

bjsowa replied:
any suggestions?

moninom1 replied:
We can make it generic, pass the argument. vSetAllNetworksUp with input parameter as xIsAllNetworksUp

bjsowa replied:
how about now?

moninom1 replied:
Yeah seems better. You might have to add Unit Test for the new function you have added for code coverage.

bjsowa replied:
I'll fix the existing unit tests and add more tests to cover new functions once the solution is accepted.

htibosch commented on source/FreeRTOS_DHCP.c:
```diff
@@ -1507,7 +1532,7 @@
         /* MISRA Ref 11.4.1 [Socket error and integer to pointer conversion] */
```
This MISRA comment can be safely removed. It was necessary earlier when the socket was compared with `FREERTOS_INVALID_SOCKET` directly. Now `xSocketValid()` is used in stead.

bjsowa replied:
done

htibosch commented on source/FreeRTOS_DHCP.c:
```diff
@@ -887,6 +909,7 @@
                 xAddress.sin_port = ( uint16_t ) dhcpCLIENT_PORT_IPv4;
```
I think it wouldn't hurt if all fields of `xAddress` will be set, as is also done in `prvCreateDHCPv6Socket()`.
~~~diff
+    memset( &xAddress, 0, sizeof( xAddress ) );
+    xAddress.sin_family = FREERTOS_AF_INET4;
+    xAddress.sin_len = ( uint8_t ) sizeof( xAddress );
     xAddress.sin_port = ( uint16_t ) dhcpCLIENT_PORT_IPv4;
~~~

bjsowa replied:
done

htibosch replied:
> done

Thank you @bjsowa for the PR. It's good work and very much appreciated. I already approved it yesterday.
Hein

htibosch reviewed (approved):
Hello @bjsowa, thank you for repairing the open/close of the DHCP client socket.

I only have two minor remarks, beside that I approve the PR.

Maybe in a future PR, we could remove the `configASSERTS()` from `prvCreateDHCPv6Socket().
They test the results of `FreeRTOS_socket()` and of `FreeRTOS_bind()`. It is not unthinkable that one of these functions will ever fail. Calling assert will just stop the application.

Thanks, Hein

bjsowa commented:
I fixed the existing unit tests and added my own for 100% coverage.

The only problem left is the CBMC proof that's failing. I would gladly appreciate someones help with it as I have no idea how to even run these tests locally.

tony-josi-aws commented:
@bjsowa, Thanks for the PR and fixing all the other checks. I can take care of fixing the CBMC proof failures.
